JSON schema for JSON blobs in SpineDB

suvayu opened the pull request:
Proposed schema for JSON blobs in SpineDB

The schema can be updated by changing the dataclasses in `models.py` and re-running the script:
```
hatch run schema/models.py schema/json-blob-schema.json
```

## Check queryability in SQL
- [ ] SQLite
- [ ] MySQL
- [ ] Postgres


## Remark
Ignore the change in `spinedbapi/`, that is only relevant for some offline testing.

suvayu committed:
JSON schema for JSON blobs in SpineDB

suvayu committed:
schema: replace parameterised types w/ type union

Pydantic doesn't understand parametrised types, use type union instead

suvayu committed:
schema: add docstring

suvayu committed:
schema: allow for arrays as index, and multiple value columns

suvayu committed:
spinedbapi: move idx_name function as free function

suvayu committed:
spinedbapi: refactor make_records to handle variations

suvayu committed:
schema: simplify model

suvayu committed:
Add a pyproject.toml for easier setup

suvayu committed:
schema: minor model definition change, add boolean as values

suvayu committed:
schema: some adhoc code to convert back and forth

suvayu committed:
schema: drop floating point numbers from indices

soininen commented:
The new schema looks good! I have some questions and comments, though.

1. Why does `Table` have an *array* of columns? My understanding is that the column contains the actual scalar values, indexed by `indices`. Why would we need more than one column?
2. What would be the use case for `Tables`?
3. The current schema seems to suggest that an index column could contain mixed data types. I think we should restrict that so that each index is of a single type. The `*Index` types could contain a `data_type` field or similar so we can validate the contents against that.
4. Values in the `*Index` types should be nullable, right? The first array in `Tables.indices` should not be nullable, though, but that we need to validate programmatically.
5. Spine time series have additional metadata: `ignore year` and `repeat` which are both booleans. Is there a way to accommodate that into the schema?
6. `Array` should accept `Duration` data type as well, but that could be just a JSON string. In the future, we may also want to add a reference data type that refers to e.g. list value or entity.
7. We use "indexes" instead of "indices" surprisingly consistently in `spinedb_api`.
8. This is really nitpicking, but I believe the `type` fields are class level constants and therefore should be annotated with `ClassVar[str]` instead of just `str`.

suvayu commented:
Thanks a lot for the in-depth review!

> * Why does `Table` have an _array_ of columns? My understanding is that the column contains the actual scalar values, indexed by `indices`. Why would we need more than one column?
> 

I was not sure on this choice.  Basically my rational for this was to support values that are more than one dimensions, e.g. GPS coordinates (lat, lon), or something else that shares the indices, say temperature, pressure, wind speed time series for the same weather station.  I don't know if this is a likely dataset for us.  If not, we can make definitely drop it.

> * What would be the use case for `Tables`?
> 

This is to support the following scenario:

2 tables in the same `parameter_value`

| bar  |       |
|------|-------|
| abc  | value |
| def  | value |

| foo  |       |
|------|-------|
| pqr  | value |
| xyz  | value |

Which we can later combine as the following.

| foo  | bar  |       |
|------|------|-------|
|      | abc  | value |
|      | def  | value |
| pqr  |      | value |
| xyz  |      | value |

I saw some examples where there was no column name, so don't know if it is okay to combine them under the same column name, so I assumed we cannot; e.g. see this simplified example [wind-node-dump-0-mini.json](https://github.com/user-attachments/files/17948764/wind-node-dump-0-mini.json).

> * The current schema seems to suggest that an index column could contain mixed data types. I think we should restrict that so that each index is of a single type. The `*Index` types could contain a `data_type` field or similar so we can validate the contents against that.
> 

Yes, I was in dilemma about this.  I was using a constrained `TypeVar` for this:

```python
T = TypeVar("T", str, int, datetime)

@dataclass
class ArrayIndex[T]:
    values: list[T]
    ...
```
But then the generated JSON would be wrong (see this commit https://github.com/ESI-FAR/Mopo-arrow_sqlite/pull/19/commits/9799ca7dcefcf03bc60578f55cc9a07e47e12f30).  I'm also not sure if JSON schema has any kind of array constraints that can do that.  The union is translated as `anyOf` which is technically correct, but not what we want.

I want to avoid additional fields that needs looking up.  So I would actually be in favour of a more verbose fallback where we define separate index arrays for all 3 supported types, and then in `Table` use a union of all the index types for the indices.  I'll also check if JSON schema supports something better than `anyOf`.

> * Values in the `*Index` types should be nullable, right? The first array in `Tables.indices` should not be nullable, though, but that we need to validate programmatically.
> 

Yes, correct.  So JSON `null` can be `None` in Python, and in PyArrow it will be interpreted as `null`.

> * Spine time series have additional metadata: `ignore year` and `repeat` which are both booleans. Is there a way to accommodate that into the schema?
> 

I didn't think about this :(.  I'll see what's possible.  I want to stay close to PyArrow function signatures, I'll experiment a bit.

> * `Array` should accept `Duration` data type as well, but that could be just a JSON string. In the future, we may also want to add a reference data type that refers to e.g. list value or entity.
> 

Yes indeed, that's how I thought about it.

> * We use "indexes" instead of "indices" surprisingly consistently in `spinedb_api`.
> 

I'll update it then.  Indexes from here on :)

> * This is really nitpicking, but I believe the `type` fields are class level constants and therefore should be annotated with `ClassVar[str]` instead of just `str`.

You are of course right.  But I think the JSON conversion doesn't understand these advanced types.  What do you think of `type: Literal["de_index"] = "de_index"`?  I don't want to heavily lean on the Python typing details because this distinction is hard to see in the JSON.  E.g. this seems to work:

```python
In [6]: @dataclass
   ...: class Foo:
   ...:     type: Literal["foo", "bar"] = "foo"
   ...:     

In [7]: RootModel[Foo].model_json_schema()
Out[7]: 
{'$defs': {'Foo': {'properties': {'type': {'default': 'foo',
     'enum': ['foo', 'bar'],
     'title': 'Type',
     'type': 'string'}},
   'title': 'Foo',
   'type': 'object'}},
 'allOf': [{'$ref': '#/$defs/Foo'}],
 'title': 'RootModel[Foo]'}
In [8]: @dataclass
   ...: class Foo:
   ...:     type: Literal["foo"] = "foo"
   ...:     

In [9]: RootModel[Foo].model_json_schema()
Out[9]: 
{'$defs': {'Foo': {'properties': {'type': {'const': 'foo',
     'default': 'foo',
     'enum': ['foo'],
     'title': 'Type',
     'type': 'string'}},
   'title': 'Foo',
   'type': 'object'}},
 'allOf': [{'$ref': '#/$defs/Foo'}],
 'title': 'RootModel[Foo]'}
```

WDYT?

soininen commented:
> I was not sure on this choice. Basically my rational for this was to support values that are more than one dimensions, e.g. GPS coordinates (lat, lon), or something else that shares the indices, say temperature, pressure, wind speed time series for the same weather station. I don't know if this is a likely dataset for us. If not, we can make definitely drop it.

I see your point. This would indeed prevent the need to repeat the indexes (especially time stamps) everywhere so data could be packed more efficiently. Also, it avoids accidental inconsistencies in the indexes. I think there is value in this, so lets leave it as it is.

> I saw some examples where there was no column name, so don't know if it is okay to combine them under the same column name, so I assumed we cannot;

There is a lot of data that does not have explicit `index_name` since we do not force it anywhere. `index_name` was actually pretty late addition to our data types.

I think the intention is to lump all data under the same column. Even though Map could be used as "struct like" data, I do not think it is used in such a way very often if at all. Even the example data in the JSON does not suggest otherwise.

> I want to avoid additional fields that needs looking up. So I would actually be in favour of a more verbose fallback where we define separate index arrays for all 3 supported types, and then in Table use a union of all the index types for the indices.

That might be nice for the schema, but I think we still need a `data_type` field in the `*Index` arrays. You need to spell the type out aloud if the array is empty in any case, and it helps a lot when working with the arrays when you know the data type before touching the data, also in other programming languages than Python.

> I didn't think about this :(. I'll see what's possible. I want to stay close to PyArrow function signatures, I'll experiment a bit.

Arrow's record batches have metadata, but I am unsure if we could use that here.

> What do you think of type: Literal["de_index"] = "de_index"?

Looks fine.

suvayu commented:
> That might be nice for the schema, but I think we still need a `data_type` field in the `*Index` arrays. You need to spell the type out aloud if the array is empty in any case, and it helps a lot when working with the arrays when you know the data type before touching the data, also in other programming languages than Python.

Okay, point taken.  Indeed, specially for other languages it might be very helpful.  Right now my thinking is reliant on PyArrow doing a lot of the type inference.  Which may not work in other contexts.

> Arrow's record batches have metadata, but I am unsure if we could use that here.

I was thinking exactly about that.  Maybe a field like `"metadata": {"post-process": {...}}`, I haven't come up with a concrete proposal yet, I've to think some more.

suvayu committed:
schema: remove fixme

suvayu committed:
Incorporate review comments from Antti

- drop Tables (batches of Table)
- add type information of values in schema

suvayu commented:
> 6\. `Array` should accept `Duration` data type as well, but that could be just a JSON string. In the future, we may also want to add a reference data type that refers to e.g. list value or entity.

Maybe we can align how we specify duration with how JSON schema [does it](https://json-schema.org/understanding-json-schema/reference/string#dates-and-times).  I'm not 100% certain this is a benefit in terms of features, so far thinking in terms of standardisation and compatibility with other tools & libraries out there.

suvayu committed:
schema: fix comment typo

suvayu committed:
schema: add duration & time pattern

- add duration & time pattern types
- restrict value types
- support encoded value columns

suvayu committed:
schema: automatic type inference of array elements

suvayu committed:
schema: update conversion to new schema

suvayu committed:
schema: add script dependency metadata (PEP 723)

suvayu committed:
schema: comment out debugging imports

suvayu committed:
schema: add missing literal

suvayu committed:
spinedbapi/insert_reencoded.py: script to insert into spinedb

suvayu committed:
spinedbapi/dbmap.py: more complete json -> records conversion

- parse 'time_series' types correctly
- handle sequence indexes during parsing

suvayu committed:
schema: add pandas types to parsed time_series

suvayu committed:
spinedbapi: update insert script

suvayu committed:
Relax dependencies to support Python 3.13

suvayu committed:
schema/models.py: reformat time-pattern regex

suvayu committed:
schema/models.py: flatten table schema

suvayu commented:
I flattened the `Table` definition to just a list of columns,  since we now have the column type as part of the schema (under `.type` and `.value_type`).

A flatter table seems simpler (I'm also thinking of potentially using JSON functions provided by the underlying db).

OleMussmann commented:
`TimePattern` pattern without `pydantic` could be something along these lines:

```python
#!/usr/bin/env python3

import re


class TimePattern:
    def __init__(self, times):
        time_pat_re = r"(Y|M|D|WD|h|m|s)[0-9]+-[0-9]+"
        pattern = rf"^{time_pat_re}[,;]{time_pat_re}$|^{time_pat_re}$"
        cpat = re.compile(pattern)

        for line in times:
            try:
                if not cpat.match(line):
                    raise self._PatternError
            except self._PatternError:
                raise TypeError(f"not following pattern: {pattern}")
            except TypeError as e:
                raise TypeError(f"not a list of strings: {e}")

        self.times = times

    class _PatternError(Exception):
        pass

    def __str__(self):
        return f"{self.__class__.__name__}(times={self.times})"


my_times = TimePattern(times=["Y2025-11", "Y2025-12"])
print(my_times)  # TimePattern(times=['Y2025-11', 'Y2025-12'])

try:
    my_wrong_times_1 = TimePattern(times=["Y2025-11", 5])
except TypeError as e:
    print(e)  # not a list of strings: ...

try:
    my_wrong_times_2 = TimePattern(times=["Y2025-11", "X2025-12"])
except TypeError as e:
    print(e)  # not following pattern: ...
```

suvayu commented:
Hi @OleMussmann, I guess it's not clear from the implementation but there are two aspects to it:
1. schema generation from the types, and
2. validation.

I'm considering (1) a priority, (2) a good to have.  Your implementation may work at runtime, but doesn't work for the schema generation.  For it to work with schema generation, it needs to be valid [annotation](https://docs.python.org/3/library/typing.html#typing.Annotated), that has the right `pydantic` hook that it generates the following schema:

```json
{
  "items": {
    "pattern": "^(Y|M|D|WD|h|m|s)[0-9]+-[0-9]+[,;](Y|M|D|WD|h|m|s)[0-9]+-[0-9]+$|^(Y|M|D|WD|h|m|s)[0-9]+-[0-9]+$",
    "type": "string"
  },
  "type": "array"
}
```

I've been unable to find the right hooks in the pydantic docs, otherwise we could adapt your implementation to make it work for schema generation.  Basically there's some "magical" dunder method we need to implement, and it will work.

OleMussmann commented:
It could be this one: https://docs.pydantic.dev/latest/concepts/types/#named-recursive-types

I'll look into it.

Commit:
add JSON examples from documentation

see: https://spine-database-api.readthedocs.io/en/stable/parameter_value_format.html

Commit:
fix parsing some data formats

OleMussmann commented:
We are merging this, because the schema is pretty stable now.  There are some pending issues, like supporting null values, and writing converter to go back to the old format.  We will deal with them in separate PRs, after reorganizing the project structure into a proper package, so that it is easier to add tests.
